### PR TITLE
Bump CAPI to v1.5.0

### PR DESCRIPTION
**Pull request by Ankitasw**
**What type of PR is this?**
/kind support

**What this PR does / why we need it**:
Bumps CAPI to v1.5.0 based on https://github.com/kubernetes-sigs/cluster-api/blob/main/docs/book/src/developer/providers/migrations/v1.4-to-v1.5.md

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes https://github.com/kubernetes-sigs/cluster-api-provider-aws/issues/4375
Fixes https://github.com/kubernetes-sigs/cluster-api-provider-aws/issues/4390

**Special notes for your reviewer**:

**Checklist**:

- [X] squashed commits
- [ ] includes documentation
- [ ] adds unit tests
- [X] adds or updates e2e tests

**Release note**:

```release-note
Bump CAPI to v1.5.0
```

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -113,7 +112,6 @@ var (
 )
 
 func main() {
-	rand.NewSource(time.Now().UnixNano())
```
Comment (cnmcavoy) [inline main.go:116]: Why is this changing? I never noticed this line before, what was this used for?

File: hack/ensure-go.sh [hunk 2]
```diff
@@ -31,7 +31,7 @@ EOF
   local go_version
   IFS=" " read -ra go_version <<< "$(go version)"
   local minimum_go_version
-  minimum_go_version=go1.20.0
+  minimum_go_version=go1.20.5
```
Comment (nrb) [inline hack/ensure-go.sh:34]: I see the triage-party image is using 1.20.6; do we want the two values to match?

File: go.mod [hunk 3]
```diff
@@ -2,7 +2,7 @@ module sigs.k8s.io/cluster-api-provider-aws/v2
 
 go 1.20
 
-replace sigs.k8s.io/cluster-api => sigs.k8s.io/cluster-api v1.4.4
+replace sigs.k8s.io/cluster-api => sigs.k8s.io/cluster-api v1.5.0
```
Comment (nrb) [inline go.mod:5]: I'm confused by this - what's the effect of replacing a package with the same thing that's required?

Comment (Ankitasw) [reply go.mod:5]: This will pick the exact version of cluster api specified, otherwise it picks up the dirty version of whatever latest CAPI main branch has.

File: hack/ensure-go.sh (hunk 2 above)
Comment (Ankitasw) [reply hack/ensure-go.sh:34]: I can change triage-party image version

File: main.go (hunk 1 above)
Comment (Ankitasw) [reply main.go:116]: Good point, I myself have not noticed this. I dont think this line has any affect, that's why it was removed by @furkatgofurov7

File: go.mod (hunk 3 above)
Comment (Ankitasw) [reply go.mod:5]: cc @richardcase to correct me if I am wrong.

File: test/e2e/data/e2e_conf.yaml [hunk 4]
```diff
@@ -169,10 +169,10 @@ variables:
   EXP_CLUSTER_RESOURCE_SET: "true"
   EXP_MACHINE_POOL: "true"
   CLUSTER_TOPOLOGY: "true"
-  INIT_WITH_BINARY_V1BETA1: "https://github.com/kubernetes-sigs/cluster-api/releases/download/v1.2.0/clusterctl-{OS}-{ARCH}"
+  INIT_WITH_BINARY_V1BETA1: "https://github.com/kubernetes-sigs/cluster-api/releases/download/v1.4.0/clusterctl-{OS}-{ARCH}"
```
Comment (Ankitasw) [inline test/e2e/data/e2e_conf.yaml:172]: Made changes to clusterctl upgrade CAPI test according to [this](https://github.com/kubernetes-sigs/cluster-api/pull/8518#issuecomment-1508064859) change, but still the tests are failing.
@killianmuldoon can you help here?
cc @richardcase

File: test/e2e/data/e2e_conf.yaml [hunk 5]
```diff
@@ -26,19 +26,30 @@ images:
...
+  - name: registry.k8s.io/cluster-api/kubeadm-control-plane-controller:v1.5.0
     loadBehavior: tryLoad
 
 providers:
   - name: cluster-api
     type: CoreProvider
     versions:
-      - name: v1.4.4 # latest published release in the v1beta1 series; this is used for v1beta1 --> main clusterctl upgrades test only.
-        value: "https://github.com/kubernetes-sigs/cluster-api/releases/download/v1.4.4/core-components.yaml"
+      - name: v1.2.0 # earliest published release in the v1beta1 series; this is used for v1beta1 old --> v1beta1 latest clusterctl upgrades test only.
```
Comment (killianmuldoon) [inline test/e2e/data/e2e_conf.yaml:40]: @Ankitasw why was this changed to v1.2.0? I think this might be the source of the issue for the upgrade tests.

**Review by killianmuldoon — COMMENTED** [review]
It looks like the test failed after creating the management cluster with CAPI v1.2. The workload cluster never gets reconciled and it looks like kubeadm fails. I don't know if the logs for the kubeadm run are collected in CAPA, but they would be useful.
- [The condition on the Cluster resource is `WaitingForKubeadmInit`](https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_cluster-api-provider-aws/4510/pull-cluster-api-provider-aws-e2e/1705606571490283520)
- [This is reflected from the `KubeadmControlPlane`](https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_cluster-api-provider-aws/4510/pull-cluster-api-provider-aws-e2e/1705606571490283520/artifacts/clusters/clusterctl-upgrade-mifc24/resources/clusterctl-upgrade/KubeadmControlPlane/clusterctl-upgrade-ms8szn-control-plane.yaml)
- The Kubernetes version is v1.27.0, but [`clusterctl init` is using CAPI v1.2.0](https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_cluster-api-provider-aws/4510/pull-cluster-api-provider-aws-e2e/1705606571490283520/artifacts/clusters/clusterctl-upgrade-mifc24/clusterctl-init.log). 
- [CAPI v1.2.0 doesn't support Kubernetes v1.27.0](https://cluster-api.sigs.k8s.io/reference/versions#kubeadm-control-plane-provider-kubeadm-control-plane-controller)

I think the first thing to try is to use a newer version of CAPI for this test - I'm not sure what the focus of this test is, but it seems more about testing the CAPA APIVersions than CAPI - or else using a Kubernetes version that's supported across versions of CAPI - v1.26.x seems to fit the bill.

File: test/e2e/data/e2e_conf.yaml (hunk 5 above)
Comment (Ankitasw) [reply test/e2e/data/e2e_conf.yaml:40]: We were using v1.4.4, I have updated this version to v1.5.0, if you see the below lines in same file, added v1.2.0 only for clusterctl upgrade tests

**Comment by Ankitasw** [comment]
> I think the first thing to try is to use a newer version of CAPI for this test - I'm not sure what the focus of this test is, but it seems more about testing the CAPA APIVersions than CAPI - or else using a Kubernetes version that's supported across versions of CAPI - v1.26.x seems to fit the bill.

Thanks for the inputs @killianmuldoon 
I have even tried using previous version of CAPI(1.4.0) with kubernetes version 1.27.0 in previous run, but even that fails with same error, so I tried reverting CAPI version for clusterctl upgrade init to v1.2.0. The current run going on is testing clusterctl init with k8s 1.26.6.

**Comment by Ankitasw** [comment]
All tests are passing.

File: .golangci.yml [hunk 6]
```diff
@@ -87,8 +87,6 @@ linters-settings:
         alias: apiextensionsv1
       - pkg: k8s.io/apimachinery/pkg/apis/meta/v1
         alias: metav1
-      - pkg: k8s.io/apimachinery/pkg/api/errors
-        alias: apierrors
```
Comment (AndiDog) [inline .golangci.yml:91]: Still used for imports, so why not lint it anymore?

**Review by AndiDog — COMMENTED** [review]
/lgtm

with minor comments

File: main.go (hunk 1 above)
Comment (Skarlso) [reply main.go:116]: Yep. The only place we are using Random is in subnets.go to shuffle the AZs. But `rand.Shuffel` is already using a default seed. So it will provide pseudo random results every time its running. Removing this should be safe.

**Comment by Skarlso** [comment]
/approve
/hold
Unhold once Andi's comments are answered. :)

File: .golangci.yml (hunk 6 above)
Comment (Ankitasw) [reply .golangci.yml:91]: This was a duplicate import in the file, so removed, otherwise we will get below error
```
ERRO [linters_context] invalid configuration, multiple aliases for the same package: pkg=k8s.io/apimachinery/pkg/api/errors aliases=[apierrors,apierrors] 
ERRO [linters_context] invalid configuration, multiple packages with the same alias: alias=apierrors packages=[k8s.io/apimachinery/pkg/api/errors,k8s.io/apimachinery/pkg/api/errors] 
```

File: pkg/cloud/services/elb/loadbalancer_test.go [hunk 7]
```diff
@@ -248,7 +248,7 @@ func TestGetAPIServerClassicELBSpecControlPlaneLoadBalancer(t *testing.T) {
 			expect: func(t *testing.T, g *WithT, res *infrav1.LoadBalancer) {
 				t.Helper()
 				expectedTarget := fmt.Sprintf("%v:%d", infrav1.ELBProtocolTCP, infrav1.DefaultAPIServerPort)
-				g.Expect(expectedTarget, res.HealthCheck.Target)
+				g.Expect(expectedTarget).NotTo(Equal(res.HealthCheck.Target))
```
[the remark this answers is not in the document]
Comment (Ankitasw) [reply pkg/cloud/services/elb/loadbalancer_test.go:251]: Great find, thanks @AndiDog yes the default target is changed when we were moving to v1beta2 APIs. This change reside in code base for long, not sure how it didn't get caught. Thanks a lot for this finding.
cc @richardcase

**Comment by Ankitasw** [comment]
Have added only the unit test change, so not re-running the E2E tests again.

**Comment by Ankitasw** [comment]
/unhold

Need one more lgtm @AndiDog @Skarlso

**Comment by Ankitasw** [comment]
Uhmm this is weird, I just changed the unit tests, maybe some changes came with rebase thats causing the issue.

**Comment by Ankitasw** [comment]
@Skarlso its fine now.
